developer_guides/topology2/topology2.rst

```diff
@@ -0,0 +1,500 @@
+	}
+
+The default type for attributes is integer. So an attribute definition with integer type can be
+defined as follows:
+
+.. code-block:: bash
+
+	DefineAttribute."index" {}
+
+Attributes definitions can contain optional constraints that are used for validating the values
+provided in the object instance. The supported type of constraints for attributes are min, max
+and valid values. The valid values array specifies the list of accepted values for the attribute.
+
+.. code-block:: bash
+
+	DefineAttribute."ramp_step_type" {
+		type	"string"
+		constraints {
+			!valid_values [
+				"linear"
+				"log"
+				"linear_zc"
+				"log_zc"
+			]
+		}
+	}
+
+Any attribute definition that contains a token_ref node is automatically added to the private
+data section of the topology object that the attribute belongs to. For example, in the above
+class definition, the attributes ramp_step_ms, ramp_step_type and UUID attributes are added
+to the private data for the widget topology object.
+
+.. code-block:: bash
+
+	DefineAttribute."ramp_step_type" {
+		type	"string"
+		# Token set reference name
+		token_ref	"sof_tkn_volume.word"
+	}
+
+Sometimes, valid_values for attributes might need to be translated from the human readable values
+to the tuple values accepted by the firmware. In the example above, valid values for ramp_step_type
+are defined as hunam reabled string values such as linera, log etc. which are translated to
+tuple values 0, 1, etc respectively before getting added to the private data section.
+
+.. code-block:: bash
+
+	DefineAttribute."ramp_step_type" {
+			type	"string"
+			# Token set reference name
+			token_ref	"sof_tkn_volume.word"
+			constraints {
+				!valid_values [
+					"linear"
+					"log"
+					"linear_zc"
+					"log_zc"
+				]
+				!tuple_values [
+					0
+					1
+					2
+					3
+				]
+			}
+		}
+
+1.1.2 Attribute qualifiers
+**************************
+Attributes in the class definition can be qualified using the following keywords:
+
+* **Constructor:** The attributes are used to construct the name of the object in the order
+  specified in the array. For example: A pga object's name based on the class definition above
+  would be constructed as "pga.1.2" where 1 and 2 are the index and instance attribute values.
+
+* **Unique:** The unique attribute is used to instantiate an object of the particular class type.
+  The attribute value must be unique when multiple objects of the same class type as instantiated
+  the same parent node. For example, when instantiating a pga object, the instance attribute must
+  have unique value when multiple PGA objects are instantiated within a pipeline.
+
+* **Mandatory:** Attributes qualified as mandatory must be provided with a value in the object
+  instance. Objects with default values in the class definition need not be qualified as mandatory.
+
+* **Immutable:** Attribute values that are set in the class definition and cannot be modified in
+  the object instance.
```

Comment:
this is a bit odd because you add the explanations *after* showing an example. I would have put this section earlier.